Lib/asyncio/timeouts.py

```diff
@@ -0,0 +1,182 @@
+        self._deadline = deadline
+
+    @property
+    def deadline(self) -> Optional[float]:
+        return self._deadline
+
+    @deadline.setter
+    def deadline(self, value: Optional[float]) -> None:
+        self._deadline = value
+        self._reschedule()
+
+    def cancel(self, msg: Any = None) -> bool:
+        return self._task.cancel(msg)
+
+    def cancelled(self) -> bool:
+        """Is timeout expired during execution?"""
+        return self._state == _State.CANCELLED
+
+    def cancelling(self) -> bool:
+        # Timeout reached, cancellation scheduled
+        return self._state in (_State.CANCELLING, _State.CANCELLED)
+
+    def __repr__(self) -> str:
+        info = [str(self._state)]
+        if self._state == _State.ENTERED and self._deadline is not None:
+            info.append("deadline={self._deadline}")
+        cls_name = self.__class__.__name__
+        return f"<{cls_name} at {id(self):#x}, {' '.join(info)}>"
+
+    def __enter__(self) -> "CancelScope":
+        self._state = _State.ENTERED
+        self._task = tasks.current_task()
+        if self._task is None:
+            raise RuntimeError("CancelScope should be used inside a task")
+        self._reschedule()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> Optional[bool]:
+        # state is EXITED if not timed out previously
+        assert self._state in (_State.ENTERED, _State.CANCELLING)
+
+        if self._timeout_handler is not None:
+            self._timeout_handler.cancel()
+            self._timeout_handler = None
+
+        if exc_type is exceptions.CancelledError:
+            if self._state == _State.CANCELLING:
+                self._state = _State.CANCELLED
+            if exc_val.args and exc_val.args[0] == id(self):
+                return True
+        elif self._state == _State.ENTERED:
+            self._state = _State.EXITED
+
+        return None
+
+    def _reschedule(self) -> None:
+        assert self._state != _State.CREATED
+        if self._state != _State.ENTERED:
+            raise RuntimeError(
+                f"Cannot change state of {self._state} CancelScope",
+            )
+
+        if self._timeout_handler is not None:
+            self._timeout_handler.cancel()
+
+        if self._deadline is None:
+            self._timeout_handler = None
+        else:
+            self._timeout_handler = self._loop.call_at(
+                self._deadline,
+                self._on_timeout,
+            )
+
+    def _on_timeout(self) -> None:
+        assert self._state == _State.ENTERED
+        self._task.cancel(id(self))
+        self._state = _State.CANCELLING
+        # drop the reference early
+        self._timeout_handler = None
+
+
+@contextlib.contextmanager
+def cancel_scope(delay: Optional[float]) -> Iterator[CancelScope]:
```

Comment:
Can one of you up with a better name for `cancel_scope()` and `cancel_scope_at()`? While I find `timeout()` and `timeout_at()` the perfect names for that API, I'm not sure that "cancel scope" is going to give anyone (who hasn't used Trio before) an idea of what it's doing. Possibly `cancel_after()` and `cancel_at()`? But possibly something else that I haven't thought of. (Agreed that `move_on()` also isn't super intuitive.)

Possibly we could just omit these two APIs and let the user catch the TimeoutError coming out of `timeout()`?

Reply:
Yeah let's just remove `move_on`.

Reply:
If we drop these apis, everything becomes close to async-timeout implementation, looks cleaner in a question what to do if two nested timeouts occur: inner converts CancelledError into TimeoutError, outer passes TimeoutError with updating the internal state to reflect that the outer timeout has also happened.

The current PR state is my attempt to design the API looking on quattro as the prime source. 
It is a good exercise, I know async-timeout pretty well and had a chance to learn an alternative approach by coding.

Reply:
dropped